hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java

```diff
@@ -107,25 +119,70 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         // it must be submitted by an account with superuser privileges, that is checked during ingest
     }
 
+    public void handle(
+            @NonNull TransactionBody txn,
+            @NonNull final AdminServiceConfig adminServiceConfig,
+            @NonNull WritableSpecialFileStore specialFileStore,
+            @NonNull SwirldDualState dualState) {
+        final FreezeUpgradeActions upgradeActions =
+                new FreezeUpgradeActions(adminServiceConfig, dualState, specialFileStore);
+
+        requireNonNull(txn);
+        requireNonNull(adminServiceConfig);
+        requireNonNull(specialFileStore);
+        // TODO: requireNonNull(dualState);
+        // for the time being, this will always be null because we are not using SwirldDualState
+
+        FreezeTransactionBody freezeTxn = txn.freeze();
+        final Timestamp freezeStartTime = freezeTxn.startTimeOrThrow();
+        final Instant freezeStartTimeInstant =
+                Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
```

Comment:
Maybe it is better to move this into the handling of only those types that require it